Tracking changes between start and end

katis opened the pull request:
We could design a pretty elegant Flutter integration with `flutter_hooks`, but it requires a way to track changes between start and end with no ability to model them as a callback.

I tried to design an API that supports the use case and hopefully prevents worst misuses,
but this should almost never be used.

Also added `dart:meta` and marked the Tracking class with an `@experimental` annotation.

See #14 for further discussion

pavanpodila commented on test/tracking.dart:
```diff
@@ -0,0 +1,57 @@
+import 'package:test/test.dart';
+
+void main() {
+  test('Tracking reacts to changes to reactive values between begin and end',
```
Can we also have a test with a autorun() to ensure the reactions are kicked off when the observables change outside

katis commented on lib/src/core/tracking.dart:
```diff
@@ -0,0 +1,40 @@
+/// This should only be used in situations where it is not possible to
+/// track changes inside a callback function.
+@experimental
+class Tracking {
```
I'll refactor context a bit, add most of the logic there, but I think this should be the public (and unsafe) API of doing tracking without callbacks. The callback API should be used for 99.9% use cases, even in Reaction & ComputedValue.

katis commented:
I changed the `Tracking` class into a `DerivationTracker`. I tried implementing the `useObserver` hook, and realized that the tracker must be usable multiple times.

The API is now:
```dart
class DerivationTracker {
    DerivationTracker(ReactiveContext context, Function() onInvalidate,
      {String name});

  void start();

  void end();
  
  void dispose();
}
```

@rrousselGit I implemented the hook like this in my experiment, does it look OK to you?

```dart
void useObserver({ReactiveContext context}) => Hook.use(context == null
    ? const _ObserverHook()
    : _ObserverHook.withContext(context));

class _ObserverHook extends Hook<void> {
  final ReactiveContext context;

  const _ObserverHook() : context = null;

  _ObserverHook.withContext(this.context) : super(keys: [context.hashCode]); // not sure about the keys

  @override
  _ObserverHookState createState() => _ObserverHookState();
}

class _ObserverHookState extends HookState<void, _ObserverHook> {
  DerivationTracker _tracker;

  @override
  void initHook() {
    super.initHook();
    _initTracker();
  }

  @override
  void didUpdateHook(_ObserverHook oldHook) {
    super.didUpdateHook(oldHook);
    _tracker?.dispose();
    _initTracker();
  }

  void _initTracker() {
    _tracker = DerivationTracker(hook.context ?? currentContext, _invalidate);
  }

  void _invalidate() => setState(noOp);

  static void noOp() {}

  @override
  void build(BuildContext context) => _tracker.start();

  @override
  void didBuild() {
    super.didBuild();
    _tracker.end();
  }

  @override
  void dispose() {
    _tracker.dispose();
    super.dispose();
  }
}
```

katis commented:
@pavanpodila could we merge this now and think about the internal refactoring of ComputedValue etc. later?

pavanpodila commented on lib/src/core/reaction.dart:
```diff
@@ -1,27 +1,39 @@
+  DerivationTracker(ReactiveContext context, Function() onInvalidate,
       {String name})
-      : super(context, onInvalidate, name: name);
+      : _reaction = Reaction(context, onInvalidate, name: name);
```
Not clear why we need a `reaction` here for tracking? Could you please elaborate? Is it mostly to wrap the `onInvalidate` in some derivation?

rrousselGit commented:
@katis There are some minor changes we can do, but that overall should do the trick, although I do not think there's a need for `didUpdateHook` at all.

pavanpodila commented:
@rrousselGit Can I add you as a collaborator? I think you bring valuable expertise that we all can benefit from?

rrousselGit commented:
Sure, please do so! 😄

pavanpodila commented:
> Sure, please do so! 😄

You are now! Please confirm on the email link from github :-). Thanks!
